### PR TITLE
vkd3d: Implement MinLODClamp using VK_EXT_image_view_min_lod

### PR DESCRIPTION
**Pull request by Joshua-Ashton**

### CODE REVIEW

File: libs/vkd3d/resource.c [hunk 1]
```diff
@@ -3481,29 +3482,37 @@ bool vkd3d_create_texture_view(struct d3d12_device *device, const struct vkd3d_t
...
-     * It will be trivial to add in RADV at least ... */
-    if (desc->miplevel_clamp >= 1.0f)
+    if (device->device_info.image_view_min_lod_features.minLod)
     {
-        uint32_t clamp_base_level;
-        uint32_t new_base_level;
-        uint32_t end_level;
+        min_lod_desc.sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_MIN_LOD_CREATE_INFO_EXT;
+        min_lod_desc.pNext = NULL;
+        min_lod_desc.minLod = desc->miplevel_clamp;
```
Comment (HansKristian-Work) [inline libs/vkd3d/resource.c:3489]: We need to handle the restriction that was added at some point during spec dev where minLod must be less-or-equal to last level accessed. We can replace the view with NULL descriptor in this case.

File: libs/vkd3d/resource.c [hunk 2]
```diff
@@ -3460,6 +3460,7 @@ static bool init_default_texture_view_desc(struct vkd3d_texture_view_desc *desc,
 bool vkd3d_create_texture_view(struct d3d12_device *device, const struct vkd3d_texture_view_desc *desc, struct vkd3d_view **view)
 {
     const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+    struct VkImageViewMinLodCreateInfoEXT min_lod_desc;
```
Comment (HansKristian-Work) [inline libs/vkd3d/resource.c:3463]: struct is not needed.

File: libs/vkd3d/resource.c (hunk 1 above)
Comment (Joshua-Ashton) [reply libs/vkd3d/resource.c]: Done,

File: libs/vkd3d/resource.c (hunk 2 above)
Comment (Joshua-Ashton) [reply libs/vkd3d/resource.c]: Done.

File: libs/vkd3d/resource.c [hunk 3]
```diff
@@ -3481,35 +3482,48 @@ bool vkd3d_create_texture_view(struct d3d12_device *device, const struct vkd3d_t
...
-        FIXME_ONCE("Cannot handle MinResourceLOD clamp of %f correctly.\n", desc->miplevel_clamp);
-
-    /* This is not correct, but it's the best we can do with existing API.
-     * It should at least avoid a scenario where implicit LOD fetches from invalid levels.
-     * TODO: We will need an extension with vkCreateImageView pNext specifying minLODClamp.
-     * It will be trivial to add in RADV at least ... */
-    if (desc->miplevel_clamp >= 1.0f)
+    /* If we clamp out of bounds, then don't make a view
+     * and use a NULL descriptor to stay in-spec. */
+    if (desc->miplevel_clamp <= (float)(desc->miplevel_count - 1))
```
Comment (HansKristian-Work) [inline libs/vkd3d/resource.c:3487]: Is miplevel_count correct? It needs to be baseMip + levelCount basically, and needs to handle VK_REMAINING_MIP_LEVELS if that's a thing.

File: libs/vkd3d/resource.c [hunk 4]
```diff
@@ -3481,35 +3482,51 @@ bool vkd3d_create_texture_view(struct d3d12_device *device, const struct vkd3d_t
...
+            min_lod_desc.pNext = NULL;
+            min_lod_desc.minLod = desc->miplevel_clamp;
+            vk_prepend_struct(&view_desc, &min_lod_desc);
         }
         else
-            view_desc.subresourceRange.baseMipLevel = clamp_base_level;
-    }
+        {
+            uint32_t clamp_base_level;
+            clamp_base_level = (uint32_t)min(desc->miplevel_clamp, (float)view_desc.subresourceRange.levelCount - 1.0f);
```
[the remark this answers is not in the document]
Comment (Joshua-Ashton) [reply libs/vkd3d/resource.c:3500]: Did we? I moved that code to here now that we don't have remaining mip layers.

[the remark this answers is not in the document]
Comment (HansKristian-Work) [reply libs/vkd3d/resource.c:3500]: We already check for `if (desc->miplevel_clamp <= (float)(desc->miplevel_idx + desc->miplevel_count - 1))`, so there is no need to further clamp miplevel_clamp.

[the remark this answers is not in the document]
Comment (Joshua-Ashton) [reply libs/vkd3d/resource.c:3500]: Oh... ye.

File: libs/vkd3d/resource.c [hunk 5]
```diff
@@ -3481,35 +3482,51 @@ bool vkd3d_create_texture_view(struct d3d12_device *device, const struct vkd3d_t
...
 
-    if ((vr = VK_CALL(vkCreateImageView(device->vk_device, &view_desc, NULL, &vk_view))) < 0)
-    {
-        WARN("Failed to create Vulkan image view, vr %d.\n", vr);
-        return false;
+            if (clamp_base_level != 0.0f)
```
[the remark this answers is not in the document]
Comment (Joshua-Ashton) [reply libs/vkd3d/resource.c]: Done.

File: libs/vkd3d/resource.c [hunk 6]
```diff
@@ -3481,35 +3482,51 @@ bool vkd3d_create_texture_view(struct d3d12_device *device, const struct vkd3d_t
...
+
+            /* This is not correct, but it's the best we can do without VK_EXT_image_view_min_lod.
+            * It should at least avoid a scenario where implicit LOD fetches from invalid levels. */
+            if (clamp_base_level >= 1.0f)
+            {
+                uint32_t new_base_level;
+                uint32_t end_level;
+
+                clamp_base_level = max((uint32_t)clamp_base_level, view_desc.subresourceRange.baseMipLevel);
+                if (view_desc.subresourceRange.levelCount != VK_REMAINING_MIP_LEVELS)
```
[the remark this answers is not in the document]
Comment (Joshua-Ashton) [reply libs/vkd3d/resource.c]: Done.

File: libs/vkd3d/resource.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (Joshua-Ashton) [reply libs/vkd3d/resource.c]: Done.

**Review by HansKristian-Work — APPROVED** [review]
LGTM

**Pull request merged**